### PR TITLE
Add Prometheus metrics for authentication attempts

### PR DESCRIPTION
**Pull request by mrogers950**
@openshift/sig-security

### CODE REVIEW

File: pkg/auth/prometheus/metrics.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	prometheus.MustRegister(authCounterTotal)
+	prometheus.MustRegister(authCounterUser)
+	prometheus.MustRegister(authCounterResult)
+	prometheus.MustRegister(authCounterUserResult)
+	prometheus.MustRegister(authCounterPath)
+	prometheus.MustRegister(authCounterUserPath)
+}
+
+func UpdateAuthCounters(user, path, result string) {
+	authCounterTotal.WithLabelValues().Inc()
```
Comment (liggitt) [inline pkg/auth/prometheus/metrics.go]: The cardinality of metrics that are labeled by username are going to be really high. What's the lifetime of these metrics?

File: pkg/auth/prometheus/metrics.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			Name:      "auth_count_path",
+			Help:      "Counts total authentication attempts, by request path",
+		}, []string{"path"},
+	)
+	authCounterUserPath = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: AuthSubsystem,
+			Name:      "auth_count_user_path",
+			Help:      "Counts total authentication attempts, by user and request path",
+		}, []string{"user", "path"},
```
Comment (smarterclayton) [inline pkg/auth/prometheus/metrics.go]: cardinality of this is too high.  Either use api request info and bucket by resource and verb, or get rid of this.

Comment (smarterclayton) [inline pkg/auth/prometheus/metrics.go]: Basically we cannot have any unbounded cardinality metrics (like path) in the system.

**Comment by mrogers950** [comment]
I removed the high cardinality label metrics.

File: pkg/auth/server/login/login.go [hunk 3]
```diff
@@ -164,19 +165,26 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
...
 	}
+	var result string
+	defer func() {
+		prometheus.UpdateAuthCounters(result)
+	}()
 	user, ok, err := l.auth.AuthenticatePassword(username, password)
 	if err != nil {
 		glog.Errorf(`Error authenticating %q with provider %q: %v`, username, l.provider, err)
 		failed(errorpage.AuthenticationErrorCode(err), w, req)
+		result = "failure"
```
Comment (liggitt) [inline pkg/auth/server/login/login.go]: Should probably distinguish this result as "error"

**Comment by mrogers950** [comment]
@liggitt @smarterclayton lmk if there are any other changes needed.

**Comment by wanghaoran1988** [comment]
@liggitt @mfojtik @smarterclayton Could you please take a look at this, this is adressing a 3.7 commited [card](https://trello.com/c/xEC43mn2/88-5-cm-ops-tools-security-prometheus-endpoint-coverage).

File: pkg/auth/prometheus/metrics.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package prometheus
```
Comment (smarterclayton) [inline pkg/auth/prometheus/metrics.go]: Call this package "metrics", we have no plans to support an alternate metrics sysem

File: pkg/auth/prometheus/metrics.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	AuthSubsystem = "auth_subsystem"
```
Comment (smarterclayton) [inline pkg/auth/prometheus/metrics.go]: Private unless there is another reason.  

Use "openshift_auth" to better describe (subsystem is redundant in public metrics

File: pkg/auth/prometheus/metrics.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		}, []string{"result"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(authCounterTotal)
+	prometheus.MustRegister(authCounterResult)
+}
+
+func UpdateAuthCounters(result string) {
```
Comment (smarterclayton) [inline pkg/auth/prometheus/metrics.go]: Generally we don't abstract metrics recording like this.  Call this method "record"

**Comment by mrogers950** [comment]
@smarterclayton updated, and rebased.

**Comment by mrogers950** [comment]
flake #15900 /retest

**Comment by mrogers950** [comment]
flake #16273

File: pkg/auth/server/login/login.go [hunk 7]
```diff
@@ -164,19 +165,26 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
...
 	}
+	var result string
+	defer func() {
+		metrics.Record(result)
+	}()
 	user, ok, err := l.auth.AuthenticatePassword(username, password)
 	if err != nil {
 		glog.Errorf(`Error authenticating %q with provider %q: %v`, username, l.provider, err)
 		failed(errorpage.AuthenticationErrorCode(err), w, req)
+		result = "error"
```
Comment (wanghaoran1988) [inline pkg/auth/server/login/login.go]: I 'd prefer the result be a constant

Comment (smarterclayton) [reply pkg/auth/server/login/login.go]: Generally we don't create constants for single use variables.

File: pkg/auth/server/login/login.go [hunk 8]
```diff
@@ -164,19 +165,26 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
 		failed(errorCodeUserRequired, w, req)
 		return
 	}
+	var result string
+	defer func() {
+		metrics.Record(result)
+	}()
 	user, ok, err := l.auth.AuthenticatePassword(username, password)
 	if err != nil {
 		glog.Errorf(`Error authenticating %q with provider %q: %v`, username, l.provider, err)
```
Comment (smarterclayton) [inline pkg/auth/server/login/login.go]: Note, this should be utilruntime.HandleError(...), not glog.Errorf().

Comment (mrogers950) [reply pkg/auth/server/login/login.go]: I opened a separate PR for that.

**Pull request merged**